docs/manpages/tracker-sql.1
.TH tracker-sql 1 "October 2014" GNU "User Commands"

.SH NAME
tracker-sql \- Use SQL to query the Tracker databases.

.SH SYNOPSIS
.nf
\fBtracker sql\fR \-q <\fIsql\fR> | \-f <\fIfile\fR>
.fi

.SH DESCRIPTION
This command allows probing of the current database. When using
commands like \fBtracker sparql\fR, the SPARQL used is translated into
SQL before being run on the database. This allows direct use of the
database using SQL avoiding the SPARQL engine entirely.

The caller can run a query two ways, either by providing a \fIfile\fR
with the query or by providing a string with the \fIsql\fR query.

The \fIfile\fR argument can be either a local path or a URI. It also
does not have to be an absolute path.

.SH OPTIONS
.TP
.B \-f, \-\-file\fR=<\fIfile\fR>
Use a \fIfile\fR with SPARQL content to query. Don't forget to end all
queries with a semicolon (;) and also to use quotes around table
names. The quotes are important because most tables are named after
ontology classes like "nfo:Document" and queries will fail without the
quotes.
.TP
.B \-q, \-\-query\fR=<\fIsql\fR>
Use a \fIsql\fR string to query the database with.

.SH EXAMPLES
.TP
Show first 10 "nfo:Document" entries where the TOC is not NULL:
.BR
.nf
$ tracker sql -q 'SELECT * FROM "nfo:Document" WHERE "nfo:tableOfContents" NOT NULL LIMIT 10;'
.fi

.SH SEE ALSO
.BR tracker-sparql (1),
.BR tracker-store (1),
.BR tracker-info (1).
.TP
.BR http://en.wikipedia.org/wiki/SQL


NEWS
NEW in 1.2.2 - 2014-09-24
=========================

The main changes since 1.2.1 are:

  * Fixes: GB#737243, API break with tracker_indexing_tree_new()
  * libtracker-miner: Fixed missing Vala bindings for new DirectoryFlags.{PRIORITY|NO_STAT} enums


NEW in 1.2.1 - 2014-09-23
=========================

The main changes since 1.2.0 are:

  * ontology: Revert nfo:hasMediaStream max cardinality change until
    Philip merges his feature branch to allow database migration and
    coping for it. This won't be in the 1.2.x releases but likely in
    1.4. and master branch.

    For packagers, skipping 1.2.0 and using 1.2.1 should avoid data
    loss by way of a reindex that may be forced when users upgrade.
    This is highly important and something we want to avoid.


NEW in 1.2.0 - 2014-09-22
=========================

The main changes since 1.0.x are:

  * New Feature: Extractor now supports eBook and Comic book formats
  * New Feature: Extractor now supports BMP files
  * New Feature: Data Provider: Now there are TrackerDataProvider and TrackerEnumerator interfaces to allow 3rd parties to supply their own data to Tracker to be indexed, for example for an online or proprietary service
  * New Feature: Extractor is now 'crash-aware' that being, we have a separate data source for content that can't be extracted to avoid repeated failures from tracker-extract. This also survives through crashes that may occur due to attempted extractions
  * build: Updated doap file (added programming language and category is now "core")
  * build: Improve firefox and thunderbird detection to avoid warnings in console
  * build: Removed all GNU_SOURCE additions in sources, use AC_USE_SYSTEM_EXTENSIONS instead
  * build: Fixed many unit tests that were not properly isolated
  * build: configure now has an --enable-minimal option to ONLY build tracker-store
  * build: Reduce vala generated source warnings from GCC
  * build: Use major version only PkgConfig files. Previously, we had $library-$major-$minor (e.g. tracker-sparql-1.2, when it should be tracker-sparql-1.0) to avoid creating unnecessary work for developers using Tracker. This release corrects this and it was planned originally.
  * build: Include COPYING*GPL files in tarball
  * build: Generate Vala bindings from GIR for libtracker-miner, libtracker-control
  * build: Improved many GIR annotations for nullable cases in libtracker-miner, libtracker-control
  * build: Avoid using rpath where possible
  * build: Make check fails to generate test reports
  * build: Remove --enable-guarantee metadata, always guarantee nie:title and nie:contentCreated
  * build: Added CC_CHECK_FLAGS_APPEND m4 macros from xinelib/systemd projects to check GCC compiler flags we can use
  * ontology: nfo:hasMediaStream no longer has a max cardinality of 1, this may mean a reindex is needed
  * libtracker-bus: Fixed a crash in query_async() where splice async operation is cancelled
  * libtracker-bus: Fixed a bug where tracker-stats showed no information with Vala 0.24+
  * libtracker-sparql: All TrackerSparqlConnection APIs throw GLib.Error due to GUnixFDList _append() API
  * libtracker-sparql: Move TRACKER_ const definitions to Tracker namespace, i.e. Tracker.DBUS_*
  * libtracker-miner: Fixed reference leak with TrackerTaskPool when using external data miners
  * libtracker-miner: Add new API tracker_decorator_fs_prepend_file() as part of 'crash-aware' feature.
  * libtracker-miner: PkgConfig file now requires libmediaart (when dependent) privately
  * libtracker-extract: Fixed broken EXIF GPS coordinate extraction
  * libtracker-extract: There is no longer an example in the examples/ directory as this library is now private
  * tracker-extract: Fixed missing TRACKER_EXTRACTORS_DIR env var that was removed at some stage
  * tracker-extract: Cleaned up signal handling, SIGALRM and SIGABRT no longer used
  * tracker-extract: [DOC] Catch more MS Office MIME types supported
  * tracker-extract: [MP3] Fixed mp3 parsing for ID3v24 or ID3v23 tags
  * tracker-extract: [ISO] Extract language, copyright and author information about ISO images using new ontology
  * tracker-extract: [ODT] Improved extraction of content so we don't stop on embedded tabs and line breaks
  * tracker-extract: [GIF] Support giflib 5.1.0 as well as earlier versions
  * tracker-stats: Now EXPRESSIONS can be used to filter stats shown
  * tracker-stats: Now only common and fallback RDF types are shown by default, --all can be used to old behaviour
  * tracker-info: Add --resource-is-iri or -i to avoid filename to URN lookups
  * tracker-info: Don't show "(null)" for nie:InformationElements with no nie:url
  * tracker-control: Added a lot of strings previsouly untranslated
  * tracker-control: Removed shorthand command line switches for less common arguments (--list-*, --pause-details)
  * tracker-control: Added --watch command line option to watch changes to the database.
  * tracker-miner-fs: By default, ignore directories with '.trackerignore' file inside
  * tracker-miner-apps: New process to mine data about applications, previously part of tracker-miner-fs. Can be disabled with --disable-miner-apps
  * tracker-miner-user-guides: New process to mine data about user guides and help content, previously part of tracker-miner-fs. Can be disabled with --disable-user-guides
  * tracker-writeback: Support GIFs
  * tracker-preferences: Removed fixed column for indexed locations, column header text was cut off
  * tracker-preferences: Don't use Stock.* deprecated since Gtk+ 3.10
  * tracker-preferences: Added suggestion to reindex OR restart depending on configuration changes applied
  * tracker-preferences: General UI clean up and modernization fixing many deprecations where old GTK+ widgets were used
  * tracker-preferences: Added 'control' tab to allow users to enable finer control of what's indexed (e.g. numbers - which are normally disabled)

src/tracker/tracker-sql.c
/*
 * Copyright (C) 2010, Nokia
 * Copyright (C) 2014, Lanedo <martyn@lanedo.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#include "config.h"

#include <string.h>
#include <locale.h>

#include <glib.h>
#include <glib/gi18n.h>

#include <libtracker-common/tracker-common.h>
#include <libtracker-data/tracker-data.h>

#include "tracker-sql.h"

#define SQL_OPTIONS_ENABLED()	  \
	(file || \
	 query)

static gchar *file;
static gchar *query;

static GOptionEntry entries[] = {
	{ "file", 'f', 0, G_OPTION_ARG_FILENAME, &file,
	  "Path to use to run a query from file",
	  "FILE",
	},
	{ "query", 'q', 0, G_OPTION_ARG_STRING, &query,
	  "SQL query",
	  "SQL",
	},
	{ NULL }
};

static int
sql_by_file (void)
{
	GError *error = NULL;
	gchar *path_in_utf8;
	gsize size;

	path_in_utf8 = g_filename_to_utf8 (file, -1, NULL, NULL, &error);
	if (error) {
		g_printerr ("%s:'%s', %s\n",
		            _("Could not get UTF-8 path from path"),
		            file,
		            error->message);
		g_error_free (error);

		return EXIT_FAILURE;
	}

	g_file_get_contents (path_in_utf8, &query, &size, &error);
	if (error) {
		g_printerr ("%s:'%s', %s\n",
		            _("Could not read file"),
		            path_in_utf8,
		            error->message);
		g_error_free (error);
		g_free (path_in_utf8);

		return EXIT_FAILURE;
	}

	g_free (path_in_utf8);

	return EXIT_SUCCESS;
}

static int
sql_by_query (void)
{
	TrackerDBInterface *iface;
	TrackerDBStatement *stmt;
	TrackerDBCursor *cursor = NULL;
	GError *error = NULL;
	gboolean first_time = FALSE;
	gint n_rows = 0;

	if (!tracker_data_manager_init (0,
	                                NULL,
	                                &first_time,
	                                FALSE,
	                                FALSE,
	                                100,
	                                100,
	                                NULL,
	                                NULL,
	                                NULL,
	                                &error)) {
		g_printerr ("%s: %s\n",
		            _("Failed to initialize data manager"),
		            error->message);
		g_error_free (error);
		return EXIT_FAILURE;
	}

	g_print ("--------------------------------------------------\n");
	g_print ("\n\n");

	iface = tracker_db_manager_get_db_interface ();

	stmt = tracker_db_interface_create_statement (iface, TRACKER_DB_STATEMENT_CACHE_TYPE_NONE, &error, "%s", query);

	if (stmt) {
		cursor = tracker_db_statement_start_cursor (stmt, &error);
	}

	if (error) {
		g_printerr ("%s: %s\n",
		            _("Could not run query"),
		            error->message);
		g_error_free (error);

		return EXIT_FAILURE;
	}

	g_print ("%s:\n", _("Results"));

	while (tracker_db_cursor_iter_next (cursor, NULL, &error)) {
		guint i;

		for (i =  0; i < tracker_db_cursor_get_n_columns (cursor); i++) {
			const gchar *str;

			if (i)
				g_print (" | ");

			str = tracker_db_cursor_get_string (cursor, i, NULL);
			if (str) {
				g_print ("%s", str);
			} else {
				g_print ("(null)");
			}
		}

		g_print ("\n");

		n_rows++;
	}

	if (error) {
		g_printerr ("%s: %s\n",
		            _("Could not run query"),
		            error->message);
		g_error_free (error);

		return EXIT_FAILURE;
	}

	if (n_rows == 0) {
		g_print ("%s\n", _("Empty result set"));
	}

	return EXIT_SUCCESS;
}

static int
sql_run (void)
{
	if (file) {
		return sql_by_file ();
	}

	if (query) {
		return sql_by_query ();
	}

	/* All known options have their own exit points */
	g_warn_if_reached ();

	return EXIT_SUCCESS;
}

static int
sql_run_default (void)
{
	GOptionContext *context;
	gchar *help;

	context = g_option_context_new (NULL);
	g_option_context_add_main_entries (context, entries, NULL);
	help = g_option_context_get_help (context, TRUE, NULL);
	g_option_context_free (context);
	g_printerr ("%s\n", help);
	g_free (help);

	return EXIT_FAILURE;
}

static gboolean
sql_options_enabled (void)
{
	return SQL_OPTIONS_ENABLED ();
}

int
tracker_sql (int argc, const char **argv)
{
	GOptionContext *context;
	GError *error = NULL;
	const gchar *failed;

	context = g_option_context_new (NULL);
	g_option_context_add_main_entries (context, entries, NULL);

	argv[0] = "tracker sql";

	if (!g_option_context_parse (context, &argc, (char***) &argv, &error)) {
		g_printerr ("%s, %s\n", _("Unrecognized options"), error->message);
		g_error_free (error);
		g_option_context_free (context);
		return EXIT_FAILURE;
	}

	g_option_context_free (context);

	if (file && query) {
		failed = _("File and query can not be used together");
	} else {
		failed = NULL;
	}

	if (failed) {
		g_printerr ("%s\n\n", failed);
		return EXIT_FAILURE;
	}

	if (sql_options_enabled ()) {
		return sql_run ();
	}

	return sql_run_default ();
}
